src/instructions.js

```diff
@@ -4841,24 +4841,26 @@ t[0xF4] = cpu => {
 
         let i = (cpu.modrm_byte >> 3 & 7) << 2;
 
-        let result = cpu.do_mul32(destination[0] , source[0]);
-        cpu.reg_xmm32s[i] = result[0];
-        cpu.reg_xmm32s[i + 1] = result[1];
+        for (let j = 0; j <= 2; j += 2)
+        {
+            let d = destination[j];
+            let s = source[j];
+            let x = i + j;
 
-        result = cpu.do_mul32(destination[2] , source[2]);
-        cpu.reg_xmm32s[i + 2] = result[0];
-        cpu.reg_xmm32s[i + 3] = result[1];
+            cpu.reg_xmm32s[x] = v86util.imul(d, s);
+            cpu.reg_xmm32s[x + 1] = (d >>> 0) * (s >>> 0) / 0x100000000;
+        }
     }
     else
     {
         // pmuludq mm1, mm2/m64
         dbg_assert((cpu.prefixes & (PREFIX_MASK_REP | PREFIX_MASK_OPSIZE)) == 0);
         let source64s = cpu.read_mmx_mem64s();
         let destination_low = cpu.reg_mmxs[2 * (cpu.modrm_byte >> 3 & 7)];
+        let source = source64s[0];
 
-        let result = cpu.do_mul32(destination_low,source64s[0])
-
-        cpu.write_mmx64s(result[0], result[1]);
+        cpu.write_mmx64s(v86util.imul(destination_low, source),
+                         (destination_low >>> 0) * (source >>> 0) / 0x100000000);
```

Comment:
Could you move this expression into a function? `v86util.imul_high` or so? (also above)

Reply:
@copy I was not sure if I should write separate versions for signed and unsigned or delegate that matter to the caller. Any feedback is appreciated.